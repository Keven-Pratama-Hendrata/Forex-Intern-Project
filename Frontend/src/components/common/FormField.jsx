import React from 'react';

const FormField = ({ 
  label, 
  type, 
  name, 
  placeholder, 
  value, 
  onChange, 
  autoComplete, 
  required = false,
  className = ''
}) => (
  <div className={`form-control w-full mb-6 space-y-1 ${className}`}>
    <label className="label">
      <span className="label-text font-medium">{label}</span>
    </label>
    <input
      type={type}
      name={name}
      placeholder={placeholder}
      className="input input-bordered w-full rounded-full"
      value={value}
      onChange={onChange}
      autoComplete={autoComplete}
      required={required}
    />
  </div>
);

export default FormField; 