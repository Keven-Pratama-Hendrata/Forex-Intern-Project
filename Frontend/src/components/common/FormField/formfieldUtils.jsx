/**
 * Creates an event handler function for form input changes that updates the form state
 * @param {Object} form Current state of the form fields as key-value pairs
 * @param {function} setForm React state setter function for updating form state
 * @returns {function} Event handler function for input change events
 */
export const handleFormChange = (form, setForm) => (e) =>
  setForm({ ...form, [e.target.name]: e.target.value });

/**
 * Validates that specified fields in a form object have non-empty values
 * @param {Object} form The form data object containing field names as keys
 * @param {Array} requiredFields Array of field names to validate
 * @param {Object} errorMessages Object mapping field names to custom error messages
 * @returns {Object} Validation result object
 * @property {boolean} isValid True if all required fields have values
 * @property {string} [message] Error message if validation fails (absent when valid)
 */
export const validateRequiredFields = (form, requiredFields, errorMessages) => {
  for (const field of requiredFields) {
    if (!form[field]) {
      return { isValid: false, message: errorMessages[field] || `${field} is required` };
    }
  }
  return { isValid: true };
};

/**
 * Extracts a user-friendly error message from API error objects
 * @param {Error} error The caught error object from API call
 * @param {string} [defaultMessage='Operation failed'] Default message when no error message exists
 * @returns {string} The error message to display to the user
 */
export const handleApiError = (error, defaultMessage = 'Operation failed') => {
  const message = error?.message || defaultMessage;
  return message;
};

/**
 * Dispatches Redux action to store authenticated user data in state
 * @param {Function} dispatch Redux store's dispatch function
 * @param {Function} loginSuccess Redux action creator that returns LOGIN_SUCCESS action
 * @param {string} token JWT authentication token received from API
 */
export const setUserData = (dispatch, loginSuccess, token) => {
  dispatch(loginSuccess({ token }));
};

/**
 * Dispatches Redux action to clear authenticated user data from state
 * @param {Function} dispatch Redux store's dispatch function
 * @param {Function} logout Redux action creator that returns LOGOUT action
 */
export const clearUserData = (dispatch, logout) => {
  dispatch(logout());
};

/**
 * Renders an input element for a form field.
 * @param {Object} props Input props.
 * @param {string} props.type Input type.
 * @param {string} props.name Input name.
 * @param {string} [props.placeholder] Placeholder text.
 * @param {string|number} [props.value] Input value.
 * @param {function} props.onChange Change handler.
 * @param {string} [props.autoComplete] Autocomplete attribute.
 * @param {boolean} [props.required] Required flag.
 * @returns {JSX.Element} Input element.
 */
export function renderInput({ type, name, placeholder, value, onChange, autoComplete, required }) {
  return (
    <input
      id={name}
      type={type}
      name={name}
      placeholder={placeholder}
      className="input input-bordered w-full rounded-full"
      value={value}
      onChange={onChange}
      autoComplete={autoComplete}
      required={required}
    />
  );
}

/**
 * Renders the wrapper for a form field with label and input.
 * @param {Object} props Wrapper props.
 * @param {string} props.label Label text.
 * @param {JSX.Element} props.input Input element.
 * @param {string} [props.className] Additional class names.
 * @param {string} props.name Input name (for htmlFor).
 * @returns {JSX.Element} Wrapper element.
 */
export function renderFormFieldWrapper({ label, input, className, name }) {
  return (
    <div className={`form-control w-full mb-6 space-y-1 ${className}`}>
      <label className="label" htmlFor={name}>
        <span className="label-text font-medium">{label}</span>
      </label>
      {input}
    </div>
  );
}