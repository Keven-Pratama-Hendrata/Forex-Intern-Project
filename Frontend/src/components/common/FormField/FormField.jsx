import React from 'react';
import { renderInput, renderFormFieldWrapper } from './formfieldUtils.jsx';

/**
 * Styled input field with label for use in forms.
 * Renders a label and a rounded input with full width and optional autocomplete and required attributes.
 *
 * @param {Object} props Props for FormField.
 * @param {string} props.label The label text displayed above the input.
 * @param {string} props.type The input type (e.g., 'text', 'password').
 * @param {string} props.name The input name attribute (used for form data and accessibility).
 * @param {string} [props.placeholder] Placeholder text for the input.
 * @param {string|number} [props.value] The current value of the input.
 * @param {function} props.onChange Callback fired when the input value changes (receives event).
 * @param {string} [props.autoComplete] Autocomplete attribute for browser autofill (e.g., 'username', 'off').
 * @param {boolean} [props.required=false] Whether the input is required for form submission.
 * @param {string} [props.className] Additional class names for the outer div.
 * @returns {JSX.Element} A labeled, styled input field for use in forms.
 */
function FormField({
  label,
  type,
  name,
  placeholder,
  value,
  onChange,
  autoComplete,
  required = false,
  className = '',
}) {
  const input = renderInput({ type, name, placeholder, value, onChange, autoComplete, required });
  return renderFormFieldWrapper({ label, input, className, name });
}

export default FormField; 