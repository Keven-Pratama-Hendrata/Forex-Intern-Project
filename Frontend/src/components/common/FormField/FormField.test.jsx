import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { describe, it, expect, jest } from '@jest/globals';
import FormField from './FormField';

describe('FormField', () => {
  it('renders with label and input', () => {
    render(<FormField label="Username" type="text" name="username" placeholder="Enter username" />);
    expect(screen.getByLabelText('Username')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Enter username')).toBeInTheDocument();
  });

  it('calls onChange when input changes', () => {
    const handleChange = jest.fn();
    render(<FormField label="Email" type="email" name="email" placeholder="Enter email" onChange={handleChange} />);
    fireEvent.change(screen.getByLabelText('Email'), { target: { value: 'abc' } });
    expect(handleChange).toHaveBeenCalled();
  });

  it('applies custom className to wrapper', () => {
    const { container } = render(
      <FormField label="Custom" type="text" name="custom" placeholder="Custom placeholder" className="custom-form-field" />
    );
    expect(container.firstChild).toHaveClass('custom-form-field');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <FormField
        label="Password"
        type="password"
        name="password"
        placeholder="Enter password"
        value=""
        onChange={() => { }}
        autoComplete="current-password"
        required
        className="test-snapshot"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
