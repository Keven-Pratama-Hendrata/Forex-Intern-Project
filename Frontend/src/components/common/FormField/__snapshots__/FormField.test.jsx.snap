// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FormField matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-control w-full mb-6 space-y-1 test-snapshot"
  >
    <label
      class="label"
      for="password"
    >
      <span
        class="label-text font-medium"
      >
        Password
      </span>
    </label>
    <input
      autocomplete="current-password"
      class="input input-bordered w-full rounded-full"
      id="password"
      name="password"
      placeholder="Enter password"
      required=""
      type="password"
      value=""
    />
  </div>
</DocumentFragment>
`;
